File: server/src/collectors/wifiNetworksWigle.js
/**
 * WiGLE WiFi Networks collector.
 * Primary: WiGLE search API (requires WIGLE_API_KEY).
 * Silent fallback when WIGLE returns nothing or the key is unset:
 *   - OpenStreetMap Overpass (public hotspots tagged internet_access=wlan)
 * The fallback needs no key and never surfaces in the source UI — it only
 * exists so the layer doesn't go empty when WIGLE is unavailable. The
 * NTT-BP and FREESPOT directory listings have no real coordinates and now
 * live in the intel pipeline (wifiHotspotsJcfw, wifiHotspotsFreespot).
 */

import { fetchOverpass } from './_liveHelpers.js';

const WIGLE_API_KEY = process.env.WIGLE_API_KEY || '';

async function tryWigleAPI() {
  if (!WIGLE_API_KEY) return null;
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10000);
    const res = await fetch(
      'https://api.wigle.net/api/v2/network/search?country=JP&latrange1=30&latrange2=45&longrange1=129&longrange2=146&resultsPerPage=100',
      {
        headers: { Authorization: `Basic ${WIGLE_API_KEY}` },
        signal: controller.signal,
      }
    );
    clearTimeout(timeout);
    if (!res.ok) return null;
    const data = await res.json();
    if (!data.results) return null;
    return data.results.map((net, i) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [net.trilong, net.trilat] },
      properties: {
        id: `WIGLE_${i}`,
        ssid: net.ssid,
        bssid: net.netid,
        encryption: net.encryption,
        channel: net.channel,
        last_seen: net.lastupdt,
        source: 'wigle_api',
      },
    }));
  } catch {
    return null;
  }
}

async function tryOSMWifiHotspots() {
  return fetchOverpass(
    'node["internet_access"="wlan"](area.jp);node["amenity"="internet_cafe"](area.jp);node["wifi"="free"](area.jp);node["internet_access"="yes"]["internet_access:fee"="no"](area.jp);',
    (el, i, coords) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: coords },
      properties: {
        id: `OSM_${el.id}`,
        ssid: el.tags?.['internet_access:ssid'] || el.tags?.name || `Wi-Fi ${i + 1}`,
        bssid: null,
        encryption: el.tags?.['internet_access:fee'] === 'no' ? 'open' : 'unknown',
        is_open: el.tags?.['internet_access:fee'] === 'no' || el.tags?.wifi === 'free',
        is_free_wifi: el.tags?.['internet_access:fee'] === 'no' || el.tags?.wifi === 'free',
        operator: el.tags?.operator || el.tags?.brand || null,
        venue: el.tags?.amenity || el.tags?.shop || 'unknown',
        zone_type: 'public_hotspot',
        country: 'JP',
        source: 'osm_overpass',
      },
    }),
  );
}

export default async function collectWifiNetworksWigle() {
  const [wigle, osm] = await Promise.allSettled([
    tryWigleAPI(),
    tryOSMWifiHotspots(),
  ]);

  const features = [];
  const liveSources = [];
  const sourceNames = ['wigle_api', 'osm_overpass'];
  for (const [i, r] of [wigle, osm].entries()) {
    if (r.status === 'fulfilled' && Array.isArray(r.value) && r.value.length > 0) {
      features.push(...r.value);
      liveSources.push(sourceNames[i]);
    }
  }

  const live = features.length > 0;
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: 'wifi_networks_wigle',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live,
      live_source: live ? liveSources.join(',') : null,
      description: 'WiGLE WiFi network search across Japan (OSM Overpass fallback when WIGLE_API_KEY unset)',
    },
  };
}

File: native/collectors/cyber/sources/wifi_networks_wigle.c
/* collectors/cyber/sources/wifi_networks_wigle.c
 * Port of server/src/collectors/wifiNetworksWigle.js — PRIMARY tryWigleAPI()
 * path. WiGLE search API, gated on WIGLE_API_KEY (0 rows when unset). The
 * silent OSM Overpass fallback (JS runs both via Promise.allSettled and
 * concatenates) is intentionally not ported (correctness-neutral). */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include <stdio.h>
#include <stdlib.h>

static void passthru(cJSON *p, const char *outk, cJSON *r, const char *ink) {
  cJSON *v = cJSON_GetObjectItem(r, ink);
  cJSON_AddItemToObject(p, outk, v ? cJSON_Duplicate(v, 1) : cJSON_CreateNull());
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *key = getenv("WIGLE_API_KEY");
  if (!key || !*key) { fprintf(stderr, "[wifi-networks-wigle] no API key\n"); return -1; }

  char auth[256];
  snprintf(auth, sizeof auth, "Authorization: Basic %s", key);
  const char *hdrs[] = { auth, NULL };
  cJSON *data = feed_get_json_h(ctx->http,
    "https://api.wigle.net/api/v2/network/search?country=JP&latrange1=30"
    "&latrange2=45&longrange1=129&longrange2=146&resultsPerPage=100",
    hdrs, 10000);
  if (!data) return -1;

  cJSON *results = cJSON_GetObjectItem(data, "results");
  if (!cJSON_IsArray(results)) { cJSON_Delete(data); return -1; }

  cJSON *features = cJSON_CreateArray();
  int i = 0;
  cJSON *net;
  cJSON_ArrayForEach(net, results) {
    cJSON *f = cJSON_CreateObject();
    cJSON_AddStringToObject(f, "type", "Feature");
    cJSON *g = cJSON_CreateObject();
    cJSON_AddStringToObject(g, "type", "Point");
    cJSON *co = cJSON_CreateArray();
    cJSON *lon = cJSON_GetObjectItem(net, "trilong");
    cJSON *lat = cJSON_GetObjectItem(net, "trilat");
    cJSON_AddItemToArray(co, lon ? cJSON_Duplicate(lon, 1) : cJSON_CreateNull());
    cJSON_AddItemToArray(co, lat ? cJSON_Duplicate(lat, 1) : cJSON_CreateNull());
    cJSON_AddItemToObject(g, "coordinates", co);
    cJSON_AddItemToObject(f, "geometry", g);

    cJSON *p = cJSON_CreateObject();                 /* EXACT JS key order */
    char id[32];
    snprintf(id, sizeof id, "WIGLE_%d", i);
    cJSON_AddStringToObject(p, "id", id);
    passthru(p, "ssid", net, "ssid");
    passthru(p, "bssid", net, "netid");
    passthru(p, "encryption", net, "encryption");
    passthru(p, "channel", net, "channel");
    passthru(p, "last_seen", net, "lastupdt");
    cJSON_AddStringToObject(p, "source", "wigle_api");
    cJSON_AddItemToObject(f, "properties", p);
    cJSON_AddItemToArray(features, f);
    i++;
  }
  cJSON_Delete(data);

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[wifi-networks-wigle] emitted %d\n", n);
  return n > 0 ? 0 : -1;
}

static const source_def wifi_networks_wigle_def = {
  .id = "wifi-networks-wigle", .collector = "cyber",
  .name = "WiGLE WiFi Networks", .name_ja = "WiGLE WiFi ネットワーク",
   .update_interval_sec = 86400, .run = run };
REGISTER_SOURCE(wifi_networks_wigle_def)
